scipy/sparse/_coo.py

```diff
@@ -666,54 +681,95 @@ def _matmul_dispatch(self, other):
-
-        try:
-            other.shape
-        except AttributeError:
-            other = other_a
-
-        if other.ndim == 1 or other.ndim == 2 and other.shape[1] == 1:
-            # dense row or column vector
-            if other.shape[0] != N:
+            if other.ndim == 1:
+                other = other.reshape((other.shape[0], 1)) # append 1 to shape
+                other_is_1d = True
+            
+            # Check if the inner dimensions match for matrix multiplication
+            if N != other.shape[-2]:
                 raise ValueError(
-                    f"{err_prefix} (n,k={N}),(k={other.shape[0]},1?)->(n,1?)"
+                    f"{err_prefix} (n,..,k={N}),(k={other.shape[-2]},..,m)->(n,..,m)"
                 )
-
-            result = self._matmul_vector(np.ravel(other))
-
-            if isinstance(other, np.matrix):
-                result = self._ascontainer(result)
-
-            if other.ndim == 2 and other.shape[1] == 1:
-                # If 'other' was an (nx1) column vector, reshape the result
-                result = result.reshape((*self.shape[:-1], 1))
-
+            
+            # If A or B has more than 2 dimensions, check for
+            # batch dimensions compatibility
+            if self.ndim > 2 or other.ndim > 2:
+                batch_shape_A = self.shape[:-2]
+                batch_shape_B = other.shape[:-2]
+                if batch_shape_A != batch_shape_B:
+                    try:
+                        # This will raise an error if the shapes are not broadcastable
+                        np.broadcast_shapes(batch_shape_A, batch_shape_B)
+                    except ValueError:
+                        raise ValueError("Batch dimensions are not broadcastable")
+            
+            result = self._matmul_sparse(other)
+
+            # reshape back if a or b were originally 1-D
+            if self_is_1d:
+                # if self was originally 1-D, reshape result accordingly
+                result = result.reshape(tuple(result.shape[:-2]) +
+                                        tuple(result.shape[-1:]))
+            if other_is_1d:
+                result = result.reshape(result.shape[:-1])
             return result
 
-        else:
-            # dense nD array or matrix ("multivector") not implemented
-            return NotImplemented
-
 
     def _matmul_multivector(self, other):
         result_dtype = upcast_char(self.dtype.char, other.dtype.char)
+        if self.ndim >= 3 or other.ndim >= 3:
+            # if self has shape (N,), reshape to (1,N)
+            self_is_1d = False
+            if self.ndim == 1:
+                self_is_1d = True
+                self = self.reshape(1, self.shape[0])
+            broadcast_shape = np.broadcast_shapes(self.shape[:-2], other.shape[:-2])
+            self_shape = broadcast_shape + self.shape[-2:]
+            other_shape = broadcast_shape + other.shape[-2:]
+            
+            self = self.broadcast_to(self_shape)
```

Comment:
Don't forget to update this call with the new name

Reply:
Done at 2f5b1b0ce4e3860e14e33d478ce24d3d51a9e447.